Enhancement/3167 implement selector all properties

danielgent opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Addresses issue #3167

## Relevant technical choices

<!-- Please describe your changes. -->

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

danielgent commented on assets/js/modules/analytics/datastore/__fixtures__/properties-ga4.json:
```diff
@@ -0,0 +1,80 @@
+[
```
Copied from https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics-4/datastore/__fixtures__/properties.json

danielgent commented on assets/js/modules/analytics/datastore/__snapshots__/properties.test.js.snap:
```diff
@@ -0,0 +1,668 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
I'm not a huge fan of snapshots but this seemed the easiest way of checking the sort order!

danielgent commented on assets/js/modules/analytics/datastore/properties.js:
```diff
@@ -407,6 +408,33 @@ const baseSelectors = {
 	} ),
+
+	/**
+	 * Gets all "classic" Universal Analytic and Google Analytics 4 (GA4) properties this account can access.
```
:eyes: my terminology here. I took this from the GA4 Google docs we have

danielgent commented on assets/js/modules/analytics/datastore/properties.js:
```diff
@@ -407,6 +408,33 @@ const baseSelectors = {
+			propertiesGA4 = select( 'modules/analytics-4' ).getProperties( accountID );
+		}
+
+		const propertiesGA4Mapped = propertiesGA4.map( ( property ) => ( { ...property, id: property._id } ) );
```
according to the fixtures (assuming these to be correct) `ua` properties have an `id` key, but `ga4` have an `_id` key

danielgent commented on assets/js/modules/analytics/datastore/properties.test.js:
```diff
@@ -255,6 +255,22 @@ describe( 'modules/analytics properties', () => {
+
+				expect( properties ).toMatchSnapshot();
+				expect( properties ).toHaveLength( 23 );
+			} );
```
I would also like to test for when the `modules/analytics-4` datastore is not available, but I don't any way to conditionally load data stores when testing

the test util registerAllStoresOn registers all the stores!
https://github.com/google/site-kit-wp/blob/4cac32bb3a3a5fc977d07546ef3fe5d7f3bc94c3/tests/js/utils.js#L359

danielgent commented on assets/js/modules/analytics/datastore/properties.test.js:
```diff
@@ -255,6 +255,22 @@ describe( 'modules/analytics properties', () => {
+
+				// Load data into stores
+				registry.dispatch( STORE_NAME ).receiveGetProperties( fixtures.propertiesProfiles.properties, { accountID } );
+				registry.dispatch( 'modules/analytics-4' ).receiveGetProperties( fixtures.propertiesGA4, { accountID } );
```
Copied store hydration code from previous tests

I could also test for when both of these selectors have not yet resolved *but* that feels out of scope for this as we are already testing that in the respective selector's tests

danielgent commented on assets/js/modules/analytics/datastore/properties.js:
```diff
@@ -407,6 +409,31 @@ const baseSelectors = {
+
+		if ( select( MODULES_ANALYTICS_4 ) ) {
+			const	propertiesGA4 = select( MODULES_ANALYTICS_4 ).getProperties( accountID );
+			const propertiesGA4Mapped = propertiesGA4.map( ( property ) => ( { ...property, id: property._id } ) );
```
according to the fixtures (assuming these to be correct) `ua` properties have an `id` key, but `ga4` have an `_id` key

eugene-manuilov commented on assets/js/modules/analytics/datastore/properties.js:
```diff
@@ -407,6 +409,31 @@ const baseSelectors = {
+			properties = properties.concat( propertiesGA4Mapped );
+		}
+
+		return sortBy( properties, [ 'name', 'id' ] );
```
@danielgent i think it won't work as expected because GA4 properties store property names in the `displayName` attribute. Let's rewrite this to use custom compare function `properties.sort( ( a, b ) => { ... } )`.

danielgent replied:
@eugene-manuilov so I can differentiate between the `ga` and `ua4` properties based on if they have the key `_id` or `id`, right? (from looking at the fixtures)

So I can do that.  :+1:   (and drop the `propertiesGA4Mapped` above)

eugene-manuilov reviewed (changes requested):
Thanks, @danielgent. It's a good start. Found a few issues that need to be addressed before we can merge this PR. Could you please take a look at it?

danielgent commented on assets/js/modules/analytics/datastore/properties.js:
```diff
@@ -407,6 +409,31 @@ const baseSelectors = {
+		let properties = select( STORE_NAME ).getProperties( accountID );
+
+		if ( select( MODULES_ANALYTICS_4 ) ) {
+			const	propertiesGA4 = select( MODULES_ANALYTICS_4 ).getProperties( accountID );
```
@eugene-manuilov I increased the size of tabs in vs-code so I can spot these tabs
![tabs-spaces](https://user-images.githubusercontent.com/8496063/116416191-70dc7e00-a83a-11eb-869f-dc1b16c9cb44.png)

I also found a setting where vscode was rewriting tabs as spaces :facepalm: 
`"editor.insertSpaces": false,`

So hopefully I'll make less of these formatting issues now

danielgent commented on assets/js/modules/analytics/datastore/properties.test.js:
```diff
@@ -255,6 +255,22 @@ describe( 'modules/analytics properties', () => {
+
+				const properties = registry.select( STORE_NAME ).getPropertiesIncludingGA4( testAccountID );
+
+				expect( properties ).toMatchSnapshot();
```
@eugene-manuilov  Makes total sense!  I can assert that each element is in the correct place, and then anyone else can quickly check that I've done the ticket correctly

danielgent commented:
@eugene-manuilov  I think I've addressed all your feedback now! Thanks for that.

There are quite a few changes so I imagine this will require another round of feedback.  All suggestions welcome :smile:

eugene-manuilov commented on assets/js/modules/analytics/datastore/properties.test.js:
```diff
@@ -255,6 +259,98 @@ describe( 'modules/analytics properties', () => {
+				expect( properties[ 3 ].displayName ).toBe( 'www.elasticpress.io' );
+			} );
+
+			it( 'returns only ua properties when ga4 datastore not available ', async () => {
```
@danielgent please take my apologies, but I think we don't need this test because it does nothing but repeats the test above. Let's remove this.

danielgent replied:
No problem @eugene-manuilov 

It just tests that the code I've written doesn't explode if the ga4 datastore isn't available.  It's not where I'd expect it to fail though :+1:

eugene-manuilov commented on assets/js/modules/analytics/datastore/properties.js:
```diff
@@ -325,6 +326,18 @@ const baseResolvers = {
 	},
 };
 
+const isGA4 = ( property ) => !! property._id;
+
+function compare( a, b ) {
+	if ( a < b ) {
+		return -1;
+	}
+	if ( a > b ) {
+		return 1;
+	}
+	return 0;
+}
```
These functions are very specific to the `getPropertiesIncludingGA4` selector and used only there. Let's move it into the selector body to encapsulate it there.

danielgent replied:
Sure thing @eugene-manuilov !  I was half expecting you to point out that they already existed in native JS or in Lodash :see_no_evil:

eugene-manuilov reviewed (changes requested):
Good work, @danielgent. There are two minor things to address and I'll approve this PR. Could you please update it?

eugene-manuilov reviewed (changes requested):
Good work, @danielgent. There are two minor things to address and I'll approve this PR. Could you please update it?

eugene-manuilov commented on assets/js/modules/analytics/datastore/properties.js:
```diff
@@ -407,6 +408,54 @@ const baseSelectors = {
+			properties = properties.concat( propertiesGA4 );
+		}
+
+		const isGA4 = ( property ) => !! property._id;
+
+		function compare( a, b ) {
+			if ( a < b ) {
+				return -1;
+			}
+			if ( a > b ) {
+				return 1;
+			}
+			return 0;
+		}
```
```suggestion
		const isGA4 = ( property ) => !! property._id;
		const compare = ( a, b ) => {
			if ( a < b ) {
				return -1;
			}
			if ( a > b ) {
				return 1;
			}
			return 0;
		}
```

eugene-manuilov commented on assets/js/modules/analytics/datastore/properties.test.js:
```diff
@@ -255,6 +256,59 @@ describe( 'modules/analytics properties', () => {
+					],
+					{ accountID }
+				);
+				registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetProperties( [
+					{
+						_id: '151753095-3',
+						_accountID: '151753095',
+						displayName: 'www.elasticpress.io',
+					},
+					{
+						_id: '151753095-4',
+						_accountID: '151753095',
+						displayName: 'troubled-tipped.example.com',
+					},
+				], { accountID } );
```
```suggestion

				registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetProperties(
					[
						{
							_id: '151753095-3',
							_accountID: '151753095',
							displayName: 'www.elasticpress.io',
						},
						{
							_id: '151753095-4',
							_accountID: '151753095',
							displayName: 'troubled-tipped.example.com',
						},
					],
					{ accountID }
				);
```

eugene-manuilov reviewed (changes requested):
Awesome! Thanks, @danielgent! Looks good to me. There are two formatting issues, I'll take care of it.

eugene-manuilov commented on assets/js/modules/analytics/datastore/properties.js:
```diff
@@ -407,6 +408,53 @@ const baseSelectors = {
+				return 1;
+			}
+			return 0;
+		}
```
Forgot to update here as well:
```suggestion
		};
```
